### PR TITLE
Drop support for EOL Python

### PR DESCRIPTION
**Pull request by hugovk**
Fixes https://github.com/jek/blinker/issues/46.

EOL Python versions are no longer supported by the core Python team and no longer receive security updates.

image

https://en.wikipedia.org/wiki/CPython#Version_history

They're also little used. Here's the pip installs for Blinker from PyPI for last month:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  50.71% |         88,260 |
| 3.6            |  28.01% |         48,749 |
| 3.5            |  14.50% |         25,235 |
| 3.4            |   6.41% |         11,153 |
| 3.3            |   0.17% |            290 |
| 3.7            |   0.13% |            234 |
| 2.6            |   0.08% |            131 |
| 3.2            |   0.00% |              2 |

Source: `pypinfo --start-date -49 --end-date -22 --percent --pip --markdown Blinker pyversion`

Dropping them eases the maintenance burden and allows more modern features of Python to be used, some of which is included here.

### CODE REVIEW

**Comment by hugovk** [comment]
Updated to add `python_requires` to help pip, and upgrade Python syntax using https://github.com/asottile/pyupgrade.

Here's pip installs for June 2018:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 3.6            |  42.43% |         69,300 |
| 2.7            |  38.80% |         63,378 |
| 3.5            |  14.65% |         23,924 |
| 3.4            |   3.72% |          6,082 |
| 3.7            |   0.29% |            468 |
| 3.3            |   0.05% |             85 |
| 2.6            |   0.05% |             74 |
| 3.2            |   0.01% |             22 |
| None           |   0.00% |              2 |
| 3.8            |   0.00% |              1 |
| Total          |         |        163,336 |

Source: `pypinfo --start-date 2018-06-01 --end-date 2018-06-30 --percent --markdown Blinker pyversion`

File: README.md [hunk 1]
```diff
@@ -9,31 +9,33 @@ interested parties to subscribe to events, or "signals".
...
+This is round two.
+Round 3!
+```
 
 See the [Blinker documentation](https://pythonhosted.org/blinker/) for more information.
 
 ## Requirements
 
-Blinker requires Python 2.4 or higher, Python 3.0 or higher, or Jython 2.5 or higher.
+Blinker requires Python 2.7, Python 3.4 or higher, or Jython 2.5 or higher.
```
Comment (ThomasWaldmann) [inline README.md]: I don't know anything about Jython, but doesn't this also need updating (Jython 2.7)?

File: docs/source/index.rst [hunk 2]
```diff
@@ -16,7 +16,7 @@ The core of Blinker is quite small but provides powerful features:
  - thread safety
 
 Blinker was written by Jason Kirtand and is provided under the MIT
-License. The library supports Python 2.4 or later; Python 3.0 or later;
+License. The library supports Python 2.7 and Python 3.4 or later;
 or Jython 2.5 or later; or PyPy 1.6 or later.
```
Comment (ThomasWaldmann) [inline docs/source/index.rst]: Jython / Pypy versions correct here?

File: setup.py [hunk 3]
```diff
@@ -17,22 +17,16 @@
       long_description=readme,
       license='MIT License',
       url='http://pythonhosted.org/blinker/',
+      python_requires='>=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*',
```
Comment (ThomasWaldmann) [inline setup.py]: `python_requires` requires `pip>=9.0`.

File: README.md (hunk 1 above)
Comment (hugovk) [reply README.md]: Yes, updated!

File: docs/source/index.rst (hunk 2 above)
Comment (hugovk) [reply docs/source/index.rst]: Updated!

File: setup.py (hunk 3 above)
Comment (hugovk) [reply setup.py]: Yes, and if the user has `pip<9.0` then it has no effect.

Comment (ThomasWaldmann) [reply setup.py]: Ah, good, no troubles then. :)

**Comment by hugovk** [comment]
Checking in: is there anything more I can do for this? Thank you!

**Comment by RDIL** [comment]
2.7 dies in a little more then a month, I think that also needs removing.

**Comment by ThomasWaldmann** [comment]
@RDIL no, I don't think 2.7 should be removed in the near future.

There might be still applications in python 2.7 which did not port / finish porting to py3 yet.

Thus, while old py3 versions can be dropped quickly, we should support py27 for a longer while still due to that.

**Comment by hugovk** [comment]
This PR is only for Python 2.4 - 2.6 and 3.0 - 3.4, all of which have low to zero download numbers.

October 2019:

| category | percent | downloads |
|----------|--------:|----------:|
| 3.6      |  37.69% |   408,481 |
| 3.7      |  29.76% |   322,536 |
| 2.7      |  23.68% |   256,637 |
| 3.5      |   4.95% |    53,644 |
| 3.4      |   1.86% |    20,187 |
| 3.8      |   1.38% |    14,949 |
| null     |   0.68% |     7,402 |
| 2.6      |   0.00% |        53 |
| 3.9      |   0.00% |         4 |
| 3.2      |   0.00% |         1 |
| 3.3      |   0.00% |         1 |
| Total    |         | 1,083,895 |

Source: `pip install -U pypistats && pypistats python_minor blinker --last-month`

**Comment by RDIL** [comment]
Thats fair

**Comment by jek** [comment]
Thanks for this. I don't have any conceptual issue with supporting older Python versions, though at some point the test infrastructure becomes difficult to maintain & it feels ok to shed a bit of this burden.

I don't have any intention of dropping 2.7 support for some time.

Python 3.8 positional-only parameters finally bring a portable way of fixing the `(*args, **kwargs)` hack on the `send()` method. If I were still actively working in Python (sadly I am not), I would look at making 3.8 the preferred native source version for Blinker and move all prior versions (2.7 and <3.8) to compatibility fallbacks.

**Pull request merged**